database/abuseemail.go

```diff
@@ -0,0 +1,180 @@
+package database
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.sia.tech/siad/build"
+)
+
+const (
+	// AbuseStatusBlocked denotes the blocked status.
+	AbuseStatusBlocked = "BLOCKED"
+
+	// AbuseStatusNotBlocked denotes the not blocked status.
+	AbuseStatusNotBlocked = "NOT BLOCKED"
+
+	// AbuseDefaultTag is the tag used when there are no tags found in the email
+	AbuseDefaultTag = "abusive"
+
+	// responseTemplateLegalNotice is a small notice we append to the response
+	// template that mentions we do not store any content on our servers
+	responseTemplateLegalNotice = `
+Please note that no content is stored on our servers, but rather on a decentralised network of hosts. 
+Therefor we are not to be held accountible for any potential abusive content they might contain.
+We will however do everything in our power to block access from said content when it gets reported.
+		
+Thank you for your reporting.
+`
+)
+
+type (
+	// AbuseEmail represent an object in the emails collection.
+	AbuseEmail struct {
+		// fields set by fetcher
+		ID        primitive.ObjectID `bson:"_id"`
+		UID       string             `bson:"email_uid"`
+		UIDRaw    uint32             `bson:"email_uid_raw"`
+		Body      []byte             `bson:"email_body"`
+		From      string             `bson:"email_from"`
+		Subject   string             `bson:"email_subject"`
+		MessageID string             `bson:"email_message_id"`
+
+		InsertedBy string    `bson:"inserted_by"`
+		InsertedAt time.Time `bson:"inserted_at"`
+
+		Skip bool `bson:"skip"`
+
+		// fields set by parser
+		ParsedAt    time.Time   `bson:"parsed_at"`
+		ParseResult AbuseReport `bson:"parse_result"`
+		Parsed      bool        `bson:"parsed"`
+
+		// fields set by blocker
+		BlockedAt   time.Time `bson:"blocked_at"`
+		BlockResult []string  `bson:"block_result"`
+		Blocked     bool      `bson:"blocked"`
+
+		// fields set by finalizer
+		Finalized   bool      `bson:"finalized"`
+		FinalizedAt time.Time `bson:"finalized_at"`
+	}
+
+	// AbuseReport contains all information about an abuse report.
+	AbuseReport struct {
+		Skylinks []string      `bson:"skylinks"`
+		Reporter AbuseReporter `bson:"reporter"`
+		Sponsor  string        `bson:"sponsor"`
+		Tags     []string      `bson:"tags"`
+	}
+
+	// AbuseReporter encapsulates some information about the reporter.
+	AbuseReporter struct {
+		Name         string `bson:"name"`
+		Email        string `bson:"email"`
+		OtherContact string `bson:"other_contact"`
+	}
+)
+
+// String returns a string representation of the abuse email
+func (a AbuseEmail) String() string {
+	// convenience variables
+	pr := a.ParseResult
+	blocked, unblocked := a.result()
+
+	var sb strings.Builder
+	sb.WriteString("\nAbuse Scanner Report:\n")
+
+	// write summary
+	sb.WriteString("\nSummary:\n")
+	if len(blocked) == 0 && len(unblocked) == 0 {
+		sb.WriteString("FAILURE - no skylinks found.\n")
+	} else if len(unblocked) == 0 {
```

Comment:
"blocked" means "successfully blocked" and "unblocked" means "failed to block", right? If not then we should probably rename those in order to make it more obvious what they mean.
```suggestion
	} else if len(unblocked) != 0 {
```

Reply:
Good catch. Probably missing a unit test.

Reply:
yeah sorry this was dumb, rewrote/moved this couple of times and ended up with a copy-paste error